common/src/main/java/io/homeassistant/companion/android/common/data/keychain/KeyChainRepositoryImpl.kt

```diff
@@ -0,0 +1,58 @@
+package io.homeassistant.companion.android.common.data.keychain
+
+import android.content.Context
+import android.security.KeyChain
+import io.homeassistant.companion.android.common.data.prefs.PrefsRepository
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import java.security.PrivateKey
+import java.security.cert.X509Certificate
+import javax.inject.Inject
+
+class KeyChainRepositoryImpl @Inject constructor(
+    private val prefsRepository: PrefsRepository
+) : KeyChainRepository {
+
+    private var alias: String? = null
+    private var key: PrivateKey? = null
+    private var chain: Array<X509Certificate>? = null
+    private var isLoading: Boolean = false
+
+    override suspend fun load(context: Context, alias: String) {
+        if (alias == null) return
+
+        this.alias = alias
+        prefsRepository.saveKeyAlias(alias)
+
+        load(context)
+    }
+
+    override suspend fun load(context: Context) = withContext(Dispatchers.IO)  {
+        if (alias == null) {
+            alias = prefsRepository.getKeyAlias()
+        }
+
+        if (alias != null && !isLoading && (key == null || chain == null)) {
+            isLoading = true
+            if (chain == null) {
+                chain = KeyChain.getCertificateChain(context, alias!!)
+            }
+            if (key == null) {
+                key = KeyChain.getPrivateKey(context, alias!!)
+            }
+            isLoading = false
+        }
```

Comment:
I'm wondering if this `isLoading` flag is the best way to handle mutual exclusion here.
If a previous loading is in progress, and there's a new `.load(...)` call, the second load request will be basically dropped. And also it's not atomic.

I think it would be better to use synchronization instead to solve this.
Suggestion: Extract the highlighted code to a new synchronized private method (annotated with `@Synchronized`).
(It's perfectly OK to block a coroutine on `Dispatchers.IO`, this dispatcher is designed for handling blocking operations)
This way every `load(...)` call would eventually lead to an actual call on `KeyChain`, so every time it's called, the caller will have up-to-date `KeyChain` information available by the time the `load` method returns.

Reply:
Should be okay, i wrapped the new synchronized `doLoad()` call inside a `runBlocking`, I hope that is what you meant.

Reply:
And yes, my code was a hack at best, I am really glad you proposed a clean solution :+1: